src/overlay/Peer.cpp

```diff
@@ -861,83 +865,101 @@ getOutboundFloodTimer(Application& app, StellarMessage const& msg)
 }
 
 void
-Peer::maybeTrimQueue()
+Peer::maybeTrimQueue(MessageType lastEnqueuedMessageType)
 {
-    auto now = mApp.getClock().now();
-    auto shouldTrim = [&](QueuedOutboundMessage const& msg) -> bool {
-        if (msg.mMessage.type() == TRANSACTION)
-        {
-            // Trim transactions that sit in the queue for too long
-            return (now - msg.mTimeEmplaced) >
-                   Peer::OUTBOUND_TRANSACTION_DROP_TIMEOUT;
-        }
-        else if (msg.mMessage.type() == SCP_MESSAGE)
-        {
-            auto slotIndex = msg.mMessage.envelope().statement.slotIndex;
-            auto minSlotToRemember =
-                mApp.getHerder().getMinLedgerSeqToRemember();
-            if (slotIndex >= minSlotToRemember &&
-                mApp.getHerder().isLatestSCPMessage(msg.mMessage))
-            {
-                return false;
-            }
-            // Trim an SCP message if it's outside of range or we have a newer
-            // statement
-            return true;
-        }
-        else
-        {
-            return false;
-        }
-    };
-
     int dropped = 0;
-    auto it = mFlowControlledQueue.cbegin();
-    while (it != mFlowControlledQueue.cend())
+    if (lastEnqueuedMessageType == TRANSACTION)
     {
-        if (shouldTrim(*it))
+        auto& queue = mOutboundQueues[1];
+
+        while (queue.size() > Peer::OUTBOUND_TRANSACTION_DROP_LIMIT)
         {
-            getOutboundFloodTimer(mApp, it->mMessage)
-                .Update(now - it->mTimeEmplaced);
-            it = mFlowControlledQueue.erase(it);
+            queue.pop_front();
             dropped++;
         }
-        else
+    }
+    else if (lastEnqueuedMessageType == SCP_MESSAGE)
+    {
+        auto& queue = mOutboundQueues[0];
+
+        // Iterate over the message queue. If we found any messages for slots we
+        // don't keep in-memory anymore, delete those. Otherwise, compare
+        // messages for the same slot and validator against the latest SCP
+        // message and drop
+        bool valueReplaced = false;
+        auto minSlotToRemember = mApp.getHerder().getMinLedgerSeqToRemember();
+
+        auto latest = queue.back();
+        bool valueReplaced = false;
+
+        for (auto it = queue.begin(); it != queue.end();)
         {
-            ++it;
+            if (it->mMessage.envelope().statement.slotIndex < minSlotToRemember)
+            {
+                it = queue.erase(it);
+                dropped++;
+            }
+            else if (mApp.getHerder().shouldReplaceSt(
```

Comment:
as this code is invoked every time a new message is placed in the "back" position, all you need to do is something like:
```
            if (it->mMessage.envelope().statement.slotIndex < minSlotToRemember)
            {
                it = queue.erase(it);
                dropped++;
            }
            else if (!valueReplaced && it != queue.end()-1 && mApp.getHerder().shouldReplaceSt(
                         it->mMessage.envelope().statement,
                         queue.back().mMessage.envelope().statement))
            {
                valueReplaced = true;
                *it = std::move(queue.back());
                queue.pop_back(); // this does not invalidate "it"
                dropped++;
                ++it;
            }
            else
            {
                ++it;
            }
```

Reply:
(invariant: you have at most 1 message per validator per slot per consensus phase)